torch/_meta_registrations.py

```diff
@@ -2695,191 +2696,527 @@ def meta_addbmm(self, batch1, batch2, *, beta=1, alpha=1):
+    torch._check(scalars.is_contiguous(), lambda: "Expected scalars to be contiguous.")
     torch._check(
-        all(isinstance(l, List) for l in [self, tensor1, tensor2])
-        and isinstance(scalars, torch.Tensor),
-        lambda: (
-            "_foreach_addc*_ op expects arguments of type: List[Tensor], List[Tensor], List[Tensor], tensor, "
-            f"but got: {type(self)}, {type(tensor1)}, {type(tensor2)}, and {type(scalars)}"
-        ),
+        scalars.ndim == 1,
+        lambda: f"Expected packed scalar Tensor to be of dimension 1. Got {scalars.ndim} instead.",
     )
-    torch._check(len(self) > 0, lambda: "input tensor list must not be empty.")
     torch._check(
-        len(self) == len(tensor1) and len(self) == len(tensor2),
-        lambda: "All input tensor lists must have the same length",
+        expect_length == scalars.shape[0],
+        lambda: (
+            f"Expected length of scalars to match input of length "
+            f"{expect_length} but got {scalars.shape[0]} instead."
+        ),
     )
+    return scalars.tolist()
+
+
+def _foreach_pointwise_op_tensor(func, inplace, op):
+    def _common(
+        input: List[Tensor],
+        tensors1: List[Tensor],
+        tensors2: List[Tensor],
+        scalars: Tensor,
+    ):
+        scalars = _convert_tensor_to_scalar_list(scalars, len(input))
+        _check_foreach_api_restrictions_pointwise_scalarlist(
+            input, tensors1, tensors2, scalars
+        )
+
+    @register_meta(func)
+    def _func_meta(
+        input: List[Tensor],
+        tensors1: List[Tensor],
+        tensors2: List[Tensor],
+        scalars: Tensor,
+    ):
+        _common(input, tensors1, tensors2)
+        return [
+            op(input[i], tensors1[i], tensors2[i], value=scalars[i])
+            for i in range(len(input))
+        ]
+
+    @register_meta(inplace)
+    def _inplace_meta(
+        input: List[Tensor],
+        tensors1: List[Tensor],
+        tensors2: List[Tensor],
+        scalars: Tensor,
+    ):
+        _common(input, tensors1, tensors2)
+        return None
+
+    return _func_meta, _inplace_meta
+
+
+def _foreach_ternary_op(func, inplace, op):
+    def _common(input: List[Tensor], tensors1: List[Tensor], tensors2: List[Tensor]):
+        _check_foreach_api_restrictions_pointwise(input, tensors1, tensors2)
+
+    @register_meta(func)
+    def _func_meta(input: List[Tensor], tensors1: List[Tensor], tensors2: List[Tensor]):
+        _common(input, tensors1, tensors2)
+        return [op(input[i], tensors1[i], tensors2[i]) for i in range(len(input))]
+
+    @register_meta(inplace)
+    def _inplace_meta(
+        input: List[Tensor], tensors1: List[Tensor], tensors2: List[Tensor]
+    ):
+        _common(input, tensors1, tensors2)
+        return None
+
+    return _func_meta, _inplace_meta
+
+
+_foreach_binary_op_list_alpha(
+    aten._foreach_add.List, aten._foreach_add_.List, torch.add
+)
```

Comment:
I apply an op like `add` here instead of using `empty` to avoid reimplementing promotion rules.